# 异常处理

## Java的异常
+ java的异常是一种class。异常可以在任何地方抛出，但只需要在最上层捕获就可以
+ java异常的继承关系：[./images/error_tree.png]
+ 异常有两个体系：Error和Exception。Error表示严重错误，程序一般无法处理。Exception则是运行
时的错误，它可以被捕获并处理
+ Exception又可以分为1）RuntimeException以及它的子类。2）非runtimeException
+ java规定，必须捕获的异常包括Exception及其子类但不包括RuntimeException
+ 捕获异常使用 try {something may raise error} catch (Exception e){do something}
+ 在方法定义的时候，使用throws Xxx表示该方法可能抛出的异常类型
+ e.printStackTrace()可以用来打印异常栈

## 捕获异常
+ finally可以保证有无错误都执行 try{} catch () {} finally {will be done}
+ finally 可以用来保证一些代码必须执行。
+ 某些情况下可以没有catch，只使用try ... finally。 比如方法本身声明了可能抛出的异常
```
void process(String file) throws IOException {
    try {
        ...
    } finally {
        System.out.println("END");
    }
}
```
+ catch捕获多种异常： try {...} catch (Exception1 | Exception 2) {...}

## 抛出异常
+ 先创建某个Exception的实例
+ 然后用throw语句抛出
```
void process2(String s) {
    if (s==null) {
        throw new NullPointerException(); //绝大部分抛出异常的代码会合并写成一行
    }
}
```

## 自定义异常
在代码中需要抛出异常时，尽量使用jdk已定义的异常。
+ 参数检查不合理，使用IllegalArgumentException
+ 自定义异常体系时，推荐从RuntimeException派生“根异常”，再派生出业务异常；
+ 自定义异常时，应该提供多种构造方法。

## 处理NullPointerException（空指针异常）
指针概念源自C语言，java本身没有指针。只有引用。当一个对象为null，我们调用其方法或其字段就会抛出NullPointerException
+ 必须明确，NullPointerException是一种代码逻辑错误，严禁使用catch来隐藏这种错误
+ 成员变量定义时初始化
+ 返回空字符串、空数组而不是null
+ java14可以启用增强异常信息看NullPointerException的位置。默认关闭，要用 -XX:+ShowCodeDetailsInExceptionMessages

## 使用断言
+ assert x>=0 : "x must >= 0";
+ Java断言的特点是：断言失败时会抛出AssertionError，导致程序结束退出。因此，断言不能用于可恢复的程序错误，只应该用于开发和测试阶段。
+ 对于可恢复的程序错误，不应该使用断言，而应该是抛出异常并在上层捕获
+ java默认关闭断言指令，遇到assert语句自动忽略，因此，应该给Java虚拟机传递 -enableassertion(-ea)
+ 实际开发中，很少使用断言。更好的方法是编写单元测试	 
